3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,171 @@
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`. By 
+#' default, a record will be considered duplicated if it shares the same 
+#' organization, site id, date, time, characteristic name, and sample fraction. 
+#' However, these options can be customized by passing a vector of column names 
+#' to the argument `duplicate_definition`.
+#' @param remove_duplicated_rows logical; should duplicated records be omitted
+#' from the cleaned dataset? Defaults to TRUE. 
 #' 
 #' @returns 
 #' Returns a formatted and harmonized data frame containing data downloaded from 
 #' the Water Quality Portal, where each row represents a unique data record.
 #' 
-clean_wqp_data <- function(wqp_data, char_names_crosswalk){
+clean_wqp_data <- function(wqp_data, char_names_crosswalk,
+                           commenttext_missing = c('analysis lost', 'not analyzed', 
+                                                   'not recorded', 'not collected', 
+                                                   'no measurement taken'),
+                           duplicate_definition = c('OrganizationIdentifier',
+                                                    'MonitoringLocationIdentifier',
+                                                    'ActivityStartDate', 
+                                                    'ActivityStartTime.Time',
+                                                    'CharacteristicName', 
+                                                    'ResultSampleFractionText'),
+                           remove_duplicated_rows = TRUE){
 
   # Clean data and assign flags if applicable
   wqp_data_clean <- wqp_data %>%
     # harmonize characteristic names by assigning a common parameter name
     # to the groups of characteristics supplied in `char_names_crosswalk`.
-    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name"))
+    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name")) %>%
+    # flag true missing results
+    flag_missing_results(., commenttext_missing) %>%
+    # flag duplicate records
+    flag_duplicates(., duplicate_definition) %>%
+    {if(remove_duplicated_rows){
+      remove_duplicates(., duplicate_definition)
+    } else {.}
+    }
+  
+  # Inform the user what we found for duplicated rows
+  if(remove_duplicated_rows){
+    message(sprintf(paste0("Removed %s duplicated records."), 
+                    nrow(wqp_data) - nrow(wqp_data_clean)))
+  }
   
   return(wqp_data_clean)
   
+}
+
+
+#' @title Flag missing results
+#' 
+#' @description 
+#' Function to flag true missing results, i.e. when the result measure value 
+#' and detection limit value are both NA, when "not reported" is found in the
+#' column "ResultDetectionConditionText", or when any of the strings from
```

Comment:
Looks like the last bit got cut off
```suggestion
#' column "ResultDetectionConditionText", or when any of the strings from
#' `commenttext_missing` are found in the column "ResultCommentText".
```